### DISCUSSION TITLE
Obtain Geometry of Elements as Polyfaces

### DISCUSSION THREAD

**Discussion by hanan-commits-work** [iTwin/itwinjs-core#3961] [Q&A]
Greetings, 
I want to create a ClipVector of a complex shaped element such as pipe.
I wanted to know a few things.

1. Referring to [2585](https://github.com/iTwin/itwinjs-core/discussions/2585#discussioncomment-3031776), I wanted to know how we can obtain geometry of elements as polyfaces. Please note that I am retrieving elementId and the Range3d of the said element. I want to know how i can get the entire geometry of the element.
2. Solution provided [here](https://github.com/iTwin/itwinjs-core/discussions/2585#discussioncomment-3101541l) by bbastings, i am trying to get an array of point3d, of all the selected elements, Please guide on how to get these points.

**Comment by bbastings** [comment]
As mentioned in the other discussion, getting the geometry as polyfaces doesn't ensure convex clip plane sets, I don't believe there is a general solution involving polyfaces and arbitrary geometric primitives. What is your plan to solve this using polyfaces?

Do you know what type of geometry is being persisted for the elements of interest? If it's SolidPrimitive Cone and TorusPipe, you could extract a center line/arc and radius to use to construct a path.

If you need to get the geometry on the frontend you can do so using [IModelConnection.Elements.loadProps](https://www.itwinjs.org/reference/core-frontend/imodelconnection/imodelconnection/imodelconnection.elements/loadprops/).

**Reply by hanan-commits-work** [reply]
Thank you. I will look into it.

**Reply by hanan-commits-work** [reply]
@bbastings , if you could tell me how you managed to obtain the point3d array, that would be helpful.
Basically, I am trying to make a path using the high, low and other corner points, but I am not getting precise result, so there is problem in my method of making those points, if you could tell me your logic, that would be great.

**Reply by bbastings** [reply]
It was just example code to show how you could construct a ClipVector from a UnionOfConvexClipPlaneSets.

Getting the points is a problem for you to solve based on whatever information you can glean from the pipes/elements of interest.

Did you try getting the geometry as I suggested to see if you have Cone/TorusPipe SolidPrimitives?

**Reply by hanan-commits-work** [reply]
Yes, I am getting Geometry using Elements.loadProps. But I am unsure how to get the SolidPrimitive type. I couldn't locate it in ElementProps.

**Comment by bbastings** [comment]
Use a GeometryStreamIterator described here: [GeometryStream](https://www.itwinjs.org/learning/common/geometrystream/) to inspect the GeometryStreamProps. A SolidPrimitive is one type of GeometryQuery.

`    const itLocal = new GeometryStreamIterator(props.geom, props.category);
    for (const entry of itLocal) {
      assertTrue(entry.primitive.type === "geometryQuery");
      assertTrue(entry.primitive.geometry instanceof SolidPrimitive);
    }`

EDIT: I should have mentioned the placement transform. If you want to iterate over the geometry in world coordinates (as opposed to local to the element's placement) you need to provide the localToWorld transform to the GeometryStreamIterator constructor, or use this convenience method: const itWorld = GeometryStreamIterator.fromGeometricElement3d(props);

NOTE: I do not know what type of geometry you actually have, SolidPrimitive is just one possibility.

**Reply by hanan-commits-work** [reply]
Hi.
So I confirmed the Geometry type, its not SolidPrimitive, but partReference. Please guide on how to proceed.

**Reply by bbastings** [reply]
You can call loadProps for the part element and iterate over it's geometry. A part is a way for geometric elements to share/instance common geometry.

`    const itWorld = GeometryStreamIterator.fromGeometricElement3d(elemProps);
    for (const entry of itWorld) {
      if ("partReference" !== entry.primitive.type)
        continue;

      const partProps = await iModel.elements.loadProps(entry.primitive.part.id, { wantGeometry: true }) as GeometryPartProps;
      assert.isDefined(partProps);

      // Iterate part geometry in context of geometric element...i.e. account for instance transform and inherited appearance...
      const itWorldPart = GeometryStreamIterator.fromGeometryPart(partProps, entry.geomParams, itWorld.partToWorld());
      for (const partEntry of itWorldPart) {
        assertTrue(partEntry.primitive.type === "geometryQuery");
	assertTrue(partEntry.primitive.geometry instanceof SolidPrimitive); }
      }
    }`

**Reply by hanan-commits-work** [reply]
Hi again,
I followed your code, and the element's primitive.type is not geometryQuery but 'brep'.

**Reply by bbastings** [reply]
Ok well, it was worth a try. There isn't a well-defined "path" you can extract from arbitrary brep geometry even if it was possible to interpret it on the front end.

**Reply by hanan-commits-work** [reply]
Hello again.
I checked the CAD model and Design Review, and it showed that the elements are of type primitives (smart solid, etc).
![MicrosoftTeams-image](https://user-images.githubusercontent.com/89520756/184297277-a26a455f-be34-4114-bdbb-494fff3e8d01.png)
But the code above showed the type to be grep of the same element.
Can you please guide why this is happening? Maybe some issue with the code?

**Reply by bbastings** [reply]
A SmartSolid stores brep geometry, so it is correct that the GeometryStream also stores the geometry as a brep.

**Reply by hanan-commits-work** [reply]
Can you please let me know what are the primitives we should add for classifiers?

**Reply by tonygem2022** [reply]
We are watching this thread with interest as this is something we are looking to emulate.

It looks though that the iTwin team are perhaps unsure how to resolve this query?

Thanks

**Reply by pmconne** [reply]
Tracking in #3883.

**Reply by tonygem2022** [reply]
Noted @pmconne.

Can you / iTwin team elaborate what this thread implies from a timeframe perspective please?

It states therein "4 tasks" on June 30 but can you point me to where it states the anticipated completion date?

I support @markschlosseratbentley point re "We should consider an alternate approach than just getting polyfaces from elements" if i understand the background properly. Having a cuboid (faces at 1/3/6/9 o'clock) would be adequate.

Thanks.

**Reply by pmconne** [reply]
@tonygem2022 you can already do a cuboid clip based on element bounding box or other criteria. That is not adequate for the original requestors, who want a more-or-less tight-fitting envelope around the actual element geometry. The four tasks were added to #3883 3 days ago as a result of investigation into what would be required to satisfy the original requests. The implementation has been tentatively scheduled for iTwin.js 3.4. The convex decomposition is the primary area of uncertainty.

**Reply by tonygem2022** [reply]
Noted @pmconne.

The current cuboid clip referred to is not exactly what we are looking for. I hope 3.4 comes out soon and this is included. By the thread, numerous others see this has core functionality as do we.

Thanks.
